### ISSUE TITLE
Strange behavior of `pm.sample()` with respect to NUTS-specific keyword arguments

### ISSUE THREAD

**Issue by mschmidt87** [pymc-devs/pymc#6009]
## Description of your problem

I noticed a strange behavior of the `pm.sample()` function in terms of specifying arguments for the NUTS sampler. As explained in the docstring, it is possible to specify `target_accept` as a direct arguments to `pm.sample()`, but for `max_treedepth` (and other arguments), this is not possible. Now, this leads to an inconsistency in the interface where `target_accept` is ignored when specified directly while `max_treedepth` is specified in a dictionary. 
I find this quite confusing because it means that specifying an argument changes the behavior of another argument. 

This is a minimal reproducer with comments explaining my observations:

**Please provide a minimal, self-contained, and reproducible example.**
```python
import numpy as np
import pymc as pm


rng = np.random.RandomState(123)

with pm.Model() as model:
    mu = pm.Normal("mu", mu=0, sigma=1)
    obs = pm.Normal("obs", mu=mu, sigma=1, observed=rng.standard_normal(100))

# Arguments passed to the NUTS sampler internally:
# target_accept = 0.99

with model:
    pm.sample(target_accept=0.99)

# Arguments passed to the NUTS sampler internally:
# max_treedepth = 15

with model:
    pm.sample(nuts={"max_treedepth": 15})

# Arguments passed to the NUTS sampler internally:
# max_treedepth = 15
# That means target_accept is not passed to the sampler but just silently dropped.

with model:
    pm.sample(nuts={"max_treedepth": 15}, target_accept=0.99)


# Arguments passed to the NUTS sampler internally:
# max_treedepth = 15, target_accept = 0.99
# Here, everything works as expected.
with model:
    pm.sample(nuts={"max_treedepth": 15, "target_accept": 0.99})
```

**Please provide any additional information below.**

## Versions and main components

* PyMC/PyMC3 Version: 4.1.3
* Aesara/Theano Version: 2.7.7
* Python Version: 3.9.11
* Operating system: Mac OS
* How did you install PyMC/PyMC3: pip

**Comment by michaelosthege** [comment]
The second case sounds like an unintended bug.

Looking at the code (https://github.com/pymc-devs/pymc/blob/73eac989fba7c74064fd1c18b5b68e84bdc80bd2/pymc/sampling.py#L530-L531) we should maybe deprecate supporting a dedicated `nuts` kwarg and instead pass everything through the `kwargs`?
Or the opposite by introducing a `nuts_kwargs` like we have `idata_kwargs`?

The hybrid approach we have right now may be the most unintuitive and fragile of all..

**Comment by mschmidt87** [comment]
> The second case sounds like an unintended bug.

Which case do you refer to exactly? I would have thought that the 3rd case (`pm.sample(nuts={"max_treedepth": 15}, target_accept=0.99)`) would be an unintended bug.

> The hybrid approach we have right now may be the most unintuitive and fragile of all..

Yes, I totally agree. For me, as an end-user, I guess both options you named would work. The problem with the first approach ("pass everything through the `kwargs`") might be that this leads to inconsistencies if I switch the step method which does not use specified kwarg?

**Comment by michaelosthege** [comment]
Yeah sorry, I meant this 👇 case as the unintended bug:
```python
# Arguments passed to the NUTS sampler internally:
# max_treedepth = 15
# That means target_accept is not passed to the sampler but just silently dropped.
with model:
    pm.sample(nuts={"max_treedepth": 15}, target_accept=0.99)
```

@pymc-devs/dev-core what do you think: Should we stop accepting `pm.sample(nuts={...})` for increased API consistency?
Options are to
1. passthrough via the regular `kwargs`
2. introduce a `nuts_kwargs` like we have `idata_kwargs`, or
3. to drop stepmethod kwarg forwarding because the user can always instantiate the step method themselves.

**Comment by ricardoV94** [comment]
Unfortunately, I would say users cannot easily initialize NUTS themselves. 

I don't think `nuts_kwargs` is better than `nuts`. 

I think we should just update `nuts` kwargs with any arguments that are provided directly to `sample` or, if anything, remove the convenience of passing them directly to `sample`. Everything would have to go in the `nuts` dict kwarg. Users might complain that it's too verbose...

**Comment by twiecki** [comment]
> I think we should just update nuts kwargs with any arguments that are provided directly to sample

I agree with this.

> if anything, remove the convenience of passing them directly to sample

I strongly disagree with this. This API is used quite a lot in all kinds of examples. And in 99% of the cases, `pm.sample()` is calling NUTS.

**Comment by mschmidt87** [comment]
If you guys are okay with it, I could make an attempt to implement this.

If the way forward is to update nuts kwargs with any arguments that are provided directly to sample, should the old way (providing e.g. `nuts = {'max_treedepth': 15}`) still be supported in the interest of backwards compatibility?

**Comment by ricardoV94** [comment]
Yes, we would just fix the bug for now, not change any API

**Comment by mschmidt87** [comment]
Okay. For fixing the bug, one would need to make sure that `target_accept` is properly updated (Fix case 3 of my example above) and also introduce a check that prevents the user from specifying `target_accept` twice, i.e., doing something like this:
`pm.sample(nuts={"max_treedepth": 15, "target_accept": 0.99}, target_accept=0.95)`, right?

**Comment by ricardoV94** [comment]
yes

**Comment by ricardoV94** [comment]
Closed via #6018
